### ISSUE TITLE
Persist state for each successful migration?

### ISSUE THREAD

**Issue by Sleavely** [Sleavely/exodus-migrations#7]
Maybe it would be a good idea to persist the state after each migration to avoid data loss if one fails?

**Comment by isaacSennerholt** [comment]
Some questions on this:
• How does one rollback without persisted state if one migration fails?
• If state persists, what happens to the persisted state if one migration fails? Automatic rollback, opt-in rollback?
• If opt-in rollback, should there be a `completeRollback` method that I can use to rollback my full history?
• What should history do? Should it only store successful runs? Or should it store all runs, no matter if they succeeded or not? I.e history of successes, or history of events?

**Comment by Sleavely** [comment]
Thanks for adding the list of things to consider. Seems like this ties in closely with how we'll handle rollbacks (#1).

If rollbacks were automated on failure we could use the in-memory state history to see which migrations managed to run in the current round, but I think it would be potentially invasive and increases difficulty in debugging failed migrations if we automatically rolled back without giving the user the chance to decide what to do.

TL;DR: Initially I think we will go for user-initiated rollbacks, which will require a persisted state for each successful migration job.

The state history should only contain successful migrations, so rolled-back and failed migrations will not be stored. This is to avoid storing information I'm not necessarily meaningful and keep our footprint small. (Otherwise we'll end up storing up-down-up-down-up events for the same migration file)  Please raise any concerns you may have with that approach because I'm basing that entirely on gut feeling.
